scripts/cluster.py

```diff
@@ -0,0 +1,131 @@
+            return
+
+    all_prop = None
+    inferences_per_summary = [set() for _ in summaries]
+    for i, summary in enumerate(summaries):
+        prop = {}
+        ranks = []
+        assert "reach-summary" in summary
+        for reach in summary["reach-summary"]:
+            assert "flow" in reach
+            assert "edges" in reach
+            flow = reach["flow"]
+            for edge in reach["edges"]:
+                assert "source" in edge
+                assert "target" in edge
+                assert "rank-0" in edge
+                rank = round(edge["rank-0"], int(settings.precision))
+                if  rank > float(settings.threshold):
+                    prop[(reach["flow"],edge["source"],edge["target"])] = len(ranks)
+                    if settings.equiv_classes:
+                        ranks.append([rank, edge["source"][0], edge["target"][0] ])
+                    else:
+                        ranks.append([rank])
+
+        if settings.equiv_classes:
+            kproto = KPrototypes(n_clusters=2, init='cao')
+            clust = kproto.fit_predict(np.matrix(ranks).A, categorical=[1,2])
+        else:
+            agg = AgglomerativeClustering(n_clusters=2, linkage="ward")
+            clust = agg.fit(ranks).labels_
+
+        fig = plt.figure()
+        ax = plt.subplot()
+
+        colors = ['green', 'red', 'blue', 'purple', 'cyan', 'orange']
+
+        means = {}
+        high = None
+        for k in set(clust):
+            kranks = [ranks[idx][0]
+                      for p,idx in prop.items()
+                      if clust[idx] == k]
+            means[k] = sum(kranks)/len(kranks)
+            if high is None or means[k] > means[high]:
+                high = k
+
+        props_to_isect = set([p for p,idx in prop.items() if clust[idx] == high])
+        inferences_per_summary[i] = props_to_isect
+        if all_prop is None:
+            all_prop = props_to_isect
+        else:
+            all_prop = all_prop.intersection(props_to_isect)
+
+        # exp_colors = ['green' for _ in ranks]
+        # for (f,s,t), idx in prop.items():
+        #     if t[0] != 'l':
+        #         exp_colors[idx] = 'red'
+        
+        #         clust_colors = [colors[l] if l >= 0 else "black" for l in clust.labels_]
+        
+        #         for k in range(0,2):
+        #             ax.scatter(clust.labels_, [r for rs in ranks for r in rs],
+        #                        c=exp_colors)
+                    
+        # fig.savefig("cluster.png")
+
+    print("Inferred", len(all_prop), "properties from", len(summaries), "summaries")
+    for p in all_prop:
+        print(" ".join(p))
+
+    print()
+    print("The following were unexpected")
+    for f, s, t in all_prop:
+        if t[0] != 'l':
+            print(f,s,t)
+
+    print("Precision: ", len([t for _,_,t in all_prop if t[0] == 'l'])/len(all_prop))
+    print("Recall: ", len([t for _,_,t in all_prop if t[0] == 'l'])/152)
```

Comment:
Could you elaborate on the constant `152`. Is this script specific to ft4, correct? If so, could you generalize the script so it works in other settings? What do the results look like?

Reply:
The only way to generalize to other settings is to also provide a "ground truth" file that records the expected properties.  Is this what you mean?

Reply:
I'll take a stab at integrating some of my policy checking code into this script.